supervisor.go

```diff
@@ -0,0 +1,123 @@
+package streams
+
+// Supervisor represents a commit supervisor.
+//
+// The Supervisor performs a commit in a concurrently-safe manner.
+// There can only ever be 1 ongoing commit at any given time.
+type Supervisor interface {
+	// Perform a global commit sequence.
+	//
+	// If triggered by a Node, that node should be passed as an argument to Commit.
+	Commit(Processor) error
+}
+
+type sourceMetadata map[Source]Metadata
+
+// Merge merges metadata from metaitems.
+//
+// TODO: Incorporate merge strategies.
+func (m sourceMetadata) Merge(items []Metaitem) sourceMetadata {
+	for _, item := range items {
+		m[item.Source] = item.Metadata.Merge(m[item.Source])
+	}
+	return m
+}
+
+// Compile-type interface check.
+var _ Supervisor = (*supervisor)(nil)
+
+type supervisor struct {
+	store Metastore
+
+	pumps map[Processor]Pump
+
+	committerMeta []Metaitem
+	markerMeta    []Metaitem
+}
+
+// NewSupervisor returns a new Supervisor instance.
+func NewSupervisor(store Metastore, pumps map[Node]Pump) Supervisor {
+	return &supervisor{
+		store: store,
+
+		pumps: remapPumps(pumps),
```

Comment:
When you create the supervisor, the pumps are not yet known/created. The interface should have a `WithPumps` function to tell it later about the pumps.

Reply:
I wanted to avoid bloating the interface with such methods - the User should not care about what they do. The task and commits will not start until the task is constructed anyway, so I don't think this is a big issue.
The one advantage to `WithPumps` that I see is that it makes it very obvious that the pumps are initialized late.

Reply:
Agreed.